### PR TITLE
CUDA Version of the Formal Integral

### PR DESCRIPTION
**Pull request by KevinCawley**
This creates the implementation of the formal_integral on a GPU using Numba's Cuda.

**Description**

There is now a file that runs the formal_integral on the GPU. If a NVIDIA GPU is detected, then the formal_integral will run using the GPU code. However, if there is no NVIDIA GPU then it will run using the numba implementation. 

This involves a new test file that tests against the numba implementation to ensure that there is no difference. 

**Motivation and context**

This is a new feature that will greatly speed up the computation of the numba integral by allowing it to run in parallel on the GPU.

**How has this been tested?**
- [ ] Testing pipeline.
- [x] Other. 

**Examples**

**Type of change**

- [ ] Bug fix. 
- [x] New feature. 
- [ ] Breaking change. 
- [ ] None of the above. 

**Checklist**

- [ ] My change requires a change to the documentation.
    - [ ] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [x] I have assigned and requested two reviewers for this pull request.

### CODE REVIEW

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 0),
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 1),
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 5),
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 6),
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 7),
+        ([0, 1, 1, 2, 3, 4, 5, 6, 6, 7, 8 , 9 , 10], 10),
+        ([10, 9, 8, 7, 6, 6, 5, 4, 3, 2, 1, 1, 0], 4), #This test case will call the else statement
+    ],
+)
+def test_line_search_cuda(nu, nu_insert):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:297]: This needs to test the line_search function between numba and cuda, not just the reverse binary search.  That should be a separate test.

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    ntest.assert_allclose(actual, expected)
+    ntest.assert_almost_equal(actual, expected)
+
+
+#Make test for every function call, even the convenience ones
+
+#Review this test with Jack!
+
+@pytest.mark.parametrize(
+    ["nu", "nu_insert"],
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:286]: `nu` should be the full line list, you can get tyhis from one of the test numba plasmas.  You should then test on reasonable frequencies the simulation would see.

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+    
+    p_loc = 0
+    if p == 1e-5:
+        p_loc = 0
+    elif p == 0.5:
+        p_loc = 1
+    elif p == 0.99:
+        p_loc = 2
+    elif p == 1:
+        p_loc = 3
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:234]: This should be a parameter in the decorator

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    expected[0] =  len(line_list_nu) - 1 - np.searchsorted(line_list_nu[::-1], nu_insert, side="right")
+
+    cuda_searchsorted_value_right_caller[1, 1](line_list_nu, nu_insert, actual)
+    
+    ntest.assert_almost_equal(actual, expected)
+
+@cuda.jit
+def cuda_searchsorted_value_right_caller(line_list_nu, nu_insert, actual):
+    x = cuda.grid(1)
+    actual[x] = len(line_list_nu) - 1 - cuda_searchsorted_value_right(line_list_nu[::-1], nu_insert)
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:322]: Should just test the searchsorted function, don't do any reversal stuff

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+    sim1 = Simulation.from_config(tardis_config)
+    sim1.run()
+    
+    basic_estimator1 = get_estimator(sim1)
+    basic_estimator2 = get_estimator(sim1)
+    
+    tau_sobolev_shape1 = basic_estimator1.Edotlu_estimator.shape
+    tau_sobolev_shape2 = basic_estimator2.Edotlu_estimator.shape
+    
+    basic_runner1 = MontecarloRunner.from_config(tardis_config, None, False)
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:384]: You only need a single runner.  In fact, you *should* use a single runner to guarantee the formal integral is acting on the same data

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+                         get_r_packet,
+                         get_numba_plasma,
+                         get_estimator) #This file is in the same test directory for easy access right now. 
+
+from tardis.montecarlo import MontecarloRunner
+from tardis.model import Radial1DModel
+from tardis.plasma.standard_plasmas import assemble_plasma
+from tardis.io.util import HDFWriterMixin
+from tardis.io.config_reader import ConfigurationError
+
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:353]: Imports go at the top

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+    p = r_inner[0] * p
+    oz = np.zeros_like(r_inner)
+    oshell_id = np.zeros_like(oz, dtype=np.int64)
+    print(dir(formal_integral_model))
+    populate_z_photosphere_caller[1, 3](r_inner, r_outer, formal_integral_model.time_explosion, p, oz, oshell_id, actual)
+    #This is to find which p is being tested, so the assert statement will not crash as
+    #actual is an array and needs to be indexed
+    p_loc = 0
+    if p == 0.0:
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:156]: See my comments below

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+@pytest.mark.parametrize("p", [0.0, 0.5, 1.0])
+def test_calculate_z_cuda(formal_integral_model, p):
+    actual = np.zeros(3)
+    inv_t = 1.0 / formal_integral_model.time_explosion
+    size = len(formal_integral_model.r_outer)
+    r_outer = formal_integral_model.r_outer 
+    for r in r_outer:
+        #This is to find which p is being tested, so the assert statement will not crash as
+        #actual is an array and needs to be indexed
+        p_loc = 0
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:111]: See comments below

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+    actual = np.zeros(4)
+    
+    h = 1.0
+    N = int(N)
+    data = np.random.random(N)
+
+    expected = formal_integral_numba.trapezoid_integration(data, h)
+    trapezoid_integration_caller[1, 4](data, h, actual)
+    
+    N_loc = 0
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:56]: See comments below

File: tardis/montecarlo/base.py [hunk 10]
```diff
@@ -5,14 +5,19 @@
...
 from scipy.special import zeta
 from tardis.montecarlo.spectrum import TARDISSpectrum
 
 from tardis.util.base import quantity_linspace
 from tardis.io.util import HDFWriterMixin
 from tardis.montecarlo import packet_source as source
-from tardis.montecarlo.montecarlo_numba.formal_integral import FormalIntegrator
+if is_available():
+    from tardis.montecarlo.montecarlo_numba.formal_integral_cuda import FormalIntegrator
+elif not is_available():
```
Comment (Rodot-) [inline tardis/montecarlo/base.py:18]: You can just do `else` here instead of calling the `is_available()` function again

Comment (KevinCawley) [reply tardis/montecarlo/base.py:18]: That would import the entire cuda library into this file, should the one function that is used just be called instead? I did the elif to make sure I would catch the false case. I didn't think there was more than two, but I can change it to just be else.

**Comment by Rodot-** [comment]
Looks like `cudatoolkit` needs to be installed from conda to run this, so this should either be in the documentation or in the environment file.  Not sure how nicely it will play with computers that don't have cuda GPUs when you install it though so I'm not sure if the environment file is the best place to put it

**Comment by epassaro** [comment]
> @epassaro it looks like the code coverage is failing from the tests, is there any way to tell it not the check there?



> @epassaro it looks like the code coverage is failing from the tests, is there any way to tell it not the check there?

It's already configured at https://github.com/tardis-sn/tardis/blob/master/.docstr.yaml

I don't know why it's not working, maybe it's a regression on newer versions. I'll check on the package repo.

**Comment by Rodot-** [comment]
> > @epassaro it looks like the code coverage is failing from the tests, is there any way to tell it not the check there?
> 
> > @epassaro it looks like the code coverage is failing from the tests, is there any way to tell it not the check there?
> 
> It's already configured at https://github.com/tardis-sn/tardis/blob/master/.docstr.yaml
> 
> I don't know why it's not working, maybe it's a regression on newer versions. I'll check on the package repo.

Thanks!

**Comment by epassaro** [comment]
Sorry, Didn't realize that parameter is commented out.

At some point, we discussed with @andrewfullard and agreed the tests should have docstrings.

**Comment by Rodot-** [comment]
> Sorry, Didn't realize that parameter is commented out.
> 
> At some point, we discussed with @andrewfullard and agreed the tests should have docstrings.

Okay, cool, good to know.  @KevinCawley looks like you need to document the tests

**Comment by Rodot-** [comment]
On the bright side, all the GPU tests are passing when I run them on a machine with an NVIDIA GPU and they all pass!

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    actual[::] = formal_integral_cuda.calculate_p_values(r, N)
+
+    ntest.assert_allclose(actual, expected, rtol=1e-14)
+
+
+@pytest.mark.skipif(
+    not GPUs_available, reason="No GPU is available to test CUDA function"
+)
+@pytest.mark.parametrize("nu_insert", np.linspace(3e12, 3e16, 10))
+def test_line_search_cuda(nu_insert, verysimple_numba_plasma):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:240]: Docstring

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+    },
+    {
+        "r": np.linspace(0, 1, 3),
+    },
+    # {"r": np.linspace(1, 2, 10, dtype=np.float64)},
+]
+
+
+@pytest.fixture(scope="function", params=TESTDATA_model)
+def formal_integral_model(request):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:117]: Docstring

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+    """
+    This calls the CUDA function and fills out
+    the array
+    """
+    x = cuda.grid(1)
+    actual[x] = formal_integral_cuda.trapezoid_integration_cuda(data, h)
+
+
+@njit(fastmath=True, error_model="numpy", parallel=False)
+def calculate_z(r, p, inv_t):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:101]: Why is this here and not in another file?

**Comment by Rodot-** [comment]
Why isn't the code coverage comparing to the current main?

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py (hunk 13 above)
Comment (KevinCawley) [reply tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:101]: I didn't want to add an import, so I just typed out the header.

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py (hunk 12 above)
Comment (KevinCawley) [reply tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:117]: This has a docstring

Comment (Rodot-) [reply tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:117]: whoops my bad

File: tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+    -------
+    float64
+    """
+    if r > p:
+        return math.sqrt(r * r - p * p) * C_INV * inv_t
+    else:
+        return 0.0
+
+
+class BoundsError(ValueError):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py:377]: ```IndexError```

File: tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py [hunk 15]
```diff
@@ -0,0 +1,514 @@
+import sys
+import numpy as np
+from astropy import units as u
+from numba import float64, int64, cuda
+import math
+
+from tardis.montecarlo.montecarlo_numba.numba_config import SIGMA_THOMSON
+
+C_INV = 3.33564e-11
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py:9]: These should be imported from tardis.constants

**Review by andrewfullard — CHANGES_REQUESTED** [review]

File: tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py (hunk 15 above)
Comment (andrewfullard) [reply tardis/montecarlo/montecarlo_numba/formal_integral_cuda.py:9]: Legacy requirement

**Review by andrewfullard — APPROVED** [review]
